### PR TITLE
bpo-40541: Add optional *counts* parameter to random.sample()

### PR DESCRIPTION
**Pull request by rhettinger**
Includes code and docs.  The tests are forthcoming.

https://bugs.python.org/issue40541

### CODE REVIEW

File: Doc/library/random.rst [hunk 1]
```diff
@@ -231,13 +231,20 @@ Functions for sequences
...
+   the probablity for the next selection.
+
    To choose a sample from a range of integers, use a :func:`range` object as an
    argument.  This is especially fast and space efficient for sampling from a large
    population:  ``sample(range(10000000), k=60)``.
 
    If the sample size is larger than the population size, a :exc:`ValueError`
    is raised.
 
+   .. versionchanged 3.9
```
Comment (remilapeyre) [inline Doc/library/random.rst]:
```suggestion
   .. versionchanged:: 3.9
```

File: Lib/random.py [hunk 2]
```diff
@@ -379,6 +383,16 @@ def sample(self, population, k):
             population = tuple(population)
         if not isinstance(population, _Sequence):
             raise TypeError("Population must be a sequence.  For dicts or sets, use sorted(d).")
+        if weights is not None:
+            cum_weights = list(_accumulate(weights))
```
Comment (remilapeyre) [inline Lib/random.py]: Hi, failure mode when `len(weights) != len(population)` is weird at the moment:

```
>>> from random import sample
>>> dealt = sample(['tens', 'low cards'], weights=[16, 32, 4], k=20)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/remi/src/cpython/Lib/random.py", line 395, in sample
    return [population[bisect(cum_weights, s)] for s in selections]
  File "/Users/remi/src/cpython/Lib/random.py", line 395, in <listcomp>
    return [population[bisect(cum_weights, s)] for s in selections]
IndexError: list index out of range
>>> dealt = sample(['tens', 'low cards'], weights=[16], k=20)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/remi/src/cpython/Lib/random.py", line 393, in sample
    selections = sample(range(total), k=k)
  File "/Users/remi/src/cpython/Lib/random.py", line 399, in sample
    raise ValueError("Sample larger than population or is negative")
ValueError: Sample larger than population or is negative
```


```suggestion
            cum_weights = list(_accumulate(weights))
            if len(cum_weights) != len(population):
                raise ValueError('The number of weights does not match the population')
```

Comment (rhettinger) [reply Lib/random.py]: Add the test. Thanks.

**Review by aeros — COMMENTED** [review]
Regarding the unit tests, would some form of non-deterministic test to ensure the weights behave as expected (across a substantial number of samples) be appropriate, along the same general theme of `test_shuffle`? 

In this case, I can imagine it would use something like `assertAlmostEqual()`, with a delta that has a very negligible chance of ever failing (assuming the behavior is correct). Something along those lines seems like it would be more precise compared to the existing tests with `assertLessEqual()`.

**Comment by rhettinger** [comment]
The core logic of sample() is already well tested, and we're reusing it here.  I prefer not to add a test that could periodically fail.

**Review by aeros — APPROVED** [review]
> The core logic of sample() is already well tested, and we're reusing it here. I prefer not to add a test that could periodically fail.

Depending on how it's written and how much flexibility the delta is given, the odds of failing would be effectively non-existent, but fair enough. I do see your point that the existing coverage for sample() is fairly extensive as it is, so it wouldn't be entirely necessary by any means.

The implementation and `test_sample_with_weights` looks fine to me, and I think the addition of a weights parameter to sample() would be very useful. Also, I experimented with the patch locally and encountered no issues. So FWIW, I'll give it my +1.

File: Lib/random.py [hunk 3]
```diff
@@ -379,8 +383,20 @@ def sample(self, population, k):
...
         n = len(population)
+        if weights is not None:
+            cum_weights = list(_accumulate(weights))
+            if len(cum_weights) != n:
+                raise ValueError('The number of weights does not match the population')
+            total = cum_weights.pop()
+            if not isinstance(total, int):
+                raise TypeError('Weights must be integers')
+            if total < 0:
+                raise ValueError('Total of weights must be greater than zero')
```
Comment (alexwlchan) [inline Lib/random.py]: Should this be `if total <= 0`?

If the weight are all zero, for example:

```python
random.sample(population=["red"], k=1, weights=[0])
```

you get a misleading error message:

```
ValueError: Sample larger than population or is negative
```

Comment (rhettinger) [reply Lib/random.py]: Okay.

**Pull request merged**